codebase/analytics-gemini/client_blackbox_test.go

```diff
@@ -0,0 +1,302 @@
+		require.Error(t, err)
+		t.Log("successfully errored as: ", err)
+	})
+
+	t.Run("fail on wrong URL", func(t *testing.T) {
+		// url in this request does not mention what scheme to use
+		// for e.g. http or https
+		url := "foo.bar"
+		cli := gemini.NewScanRepoClient(url, &http.Client{}, "", nil)
+
+		req := gemini.NewScanRepoRequest(repoURL, email)
+
+		ctx, err := testjwt.NewJWTContext("bcdd0b29-123d-11e8-a8bc-b69930b94f5c", "../../test/jwt/private_key.pem")
+		require.NoError(t, err)
+
+		// then
+		err = cli.Register(ctx, req)
+		require.Error(t, err)
+		t.Log("successfully errored as: ", err)
+	})
+
+	t.Run("fail to unmarshal response", func(t *testing.T) {
+		// server responds 200 but empty body which is not
+		// json format, so the unmarshaller will complain
+		r, err := testrecorder.New(
+			"../../test/data/gemini-scan/register-weird-response",
+			testrecorder.WithJWTMatcher("../../test/jwt/public_key.pem"),
+		)
+		require.NoError(t, err)
+		defer r.Stop()
+
+		httpClient := &http.Client{
+			Transport: r.Transport,
+		}
+
+		cli := gemini.NewScanRepoClient(url, httpClient, "", nil)
+
+		req := gemini.NewScanRepoRequest(repoURL, email)
+		ctx, err := testjwt.NewJWTContext("bcdd0b29-123d-11e8-a8bc-b69930b94f5c", "../../test/jwt/private_key.pem")
+		require.NoError(t, err)
+
+		// then
+		err = cli.Register(ctx, req)
+		require.Error(t, err)
+		t.Log("successfully errored as: ", err)
+	})
+
+	t.Run("unknown response from the server", func(t *testing.T) {
+		// given
+		// server responds with 500 here
+		r, err := testrecorder.New(
+			"../../test/data/gemini-scan/register-unknown-response",
+			testrecorder.WithJWTMatcher("../../test/jwt/public_key.pem"),
+		)
+		require.NoError(t, err)
+		defer r.Stop()
+
+		httpClient := &http.Client{
+			Transport: r.Transport,
+		}
+
+		cli := gemini.NewScanRepoClient(url, httpClient, "", nil)
+
+		req := gemini.NewScanRepoRequest(repoURL, email)
+		ctx, err := testjwt.NewJWTContext("bcdd0b29-123d-11e8-a8bc-b69930b94f5c", "../../test/jwt/private_key.pem")
+		require.NoError(t, err)
+
+		// then
+		err = cli.Register(ctx, req)
+		require.Error(t, err)
+		t.Log("successfully errored as: ", err)
+	})
+
+	t.Run("unknown response on 200", func(t *testing.T) {
+		// given
+		// the server returns 200 but some other output
+		r, err := testrecorder.New(
+			"../../test/data/gemini-scan/register-unknown-response-200",
+			testrecorder.WithJWTMatcher("../../test/jwt/public_key.pem"),
+		)
+		require.NoError(t, err)
+		defer r.Stop()
+
+		httpClient := &http.Client{
+			Transport: r.Transport,
+		}
+
+		cli := gemini.NewScanRepoClient(url, httpClient, "", nil)
+
+		req := gemini.NewScanRepoRequest(repoURL, email)
+		ctx, err := testjwt.NewJWTContext("bcdd0b29-123d-11e8-a8bc-b69930b94f5c", "../../test/jwt/private_key.pem")
+		require.NoError(t, err)
+
+		err = cli.Register(ctx, req)
+		require.Error(t, err)
+		t.Log("successfully errored as: ", err)
+	})
+}
+
+func TestDeRegister(t *testing.T) {
```

Comment:
Add test to verify logic around only deregister if 'your the last active codebase'

Reply:
this test case takes care of it

Reply:
https://github.com/fabric8-services/fabric8-wit/pull/2144/files#diff-0ab0db37537b51dae4485c5435e3af12R186

Reply:
Only if 1 space has the codebase. What if 2 has it imported?

Reply:
the code to check if there are codebases are active is changed, so that is unit tested.

Reply:
and even if there are codebases across spaces, it will show up in the query of 

```bash
export url=https://github.com/surajssd/heyheyworld 
curl  http://minishift.local:30000/api/search/codebases?url="$url"
```

output :

<details>

```json
{
  "data": [
    {
      "attributes": {
        "createdAt": "2018-07-09T10:39:52.393755Z",
        "cve-scan": true,
        "last_used_workspace": "",
        "type": "git",
        "url": "https://github.com/surajssd/heyheyworld"
      },
      "id": "fc2caa2e-e978-40de-bc77-0c92528b6b3f",
      "links": {
        "edit": "http://minishift.local:30000/api/codebases/fc2caa2e-e978-40de-bc77-0c92528b6b3f/edit",
        "related": "http://minishift.local:30000/api/codebases/fc2caa2e-e978-40de-bc77-0c92528b6b3f",
        "self": "http://minishift.local:30000/api/codebases/fc2caa2e-e978-40de-bc77-0c92528b6b3f"
      },
      "relationships": {
        "space": {
          "data": {
            "id": "1b6be181-7f67-4e9b-a6f4-8588565df4bd",
            "type": "spaces"
          },
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd",
            "self": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd"
          }
        },
        "workspaces": {
          "links": {
            "related": "http://minishift.local:30000/api/codebases/fc2caa2e-e978-40de-bc77-0c92528b6b3f/workspaces",
            "self": "http://minishift.local:30000/api/codebases/fc2caa2e-e978-40de-bc77-0c92528b6b3f/workspaces"
          }
        }
      },
      "type": "codebases"
    },
    {
      "attributes": {
        "createdAt": "2018-07-09T10:38:07.766919Z",
        "cve-scan": true,
        "last_used_workspace": "",
        "type": "git",
        "url": "https://github.com/surajssd/heyheyworld"
      },
      "id": "0e7407b8-6e3f-40b2-a1f5-75f0c218dafa",
      "links": {
        "edit": "http://minishift.local:30000/api/codebases/0e7407b8-6e3f-40b2-a1f5-75f0c218dafa/edit",
        "related": "http://minishift.local:30000/api/codebases/0e7407b8-6e3f-40b2-a1f5-75f0c218dafa",
        "self": "http://minishift.local:30000/api/codebases/0e7407b8-6e3f-40b2-a1f5-75f0c218dafa"
      },
      "relationships": {
        "space": {
          "data": {
            "id": "a36c3667-2877-4ef0-a745-524917fa6138",
            "type": "spaces"
          },
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138",
            "self": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138"
          }
        },
        "workspaces": {
          "links": {
            "related": "http://minishift.local:30000/api/codebases/0e7407b8-6e3f-40b2-a1f5-75f0c218dafa/workspaces",
            "self": "http://minishift.local:30000/api/codebases/0e7407b8-6e3f-40b2-a1f5-75f0c218dafa/workspaces"
          }
        }
      },
      "type": "codebases"
    }
  ],
  "included": [
    {
      "attributes": {
        "created-at": "2018-07-09T10:36:56.591364Z",
        "description": "This is the foobar collaboration space",
        "name": "space_1",
        "updated-at": "2018-07-09T10:36:56.591364Z",
        "version": 0
      },
      "id": "a36c3667-2877-4ef0-a745-524917fa6138",
      "links": {
        "backlog": {
          "self": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/backlog"
        },
        "filters": "http://minishift.local:30000/api/filters",
        "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138",
        "self": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138",
        "workitemlinktypes": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemlinktypes",
        "workitemtypes": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypes"
      },
      "relationships": {
        "areas": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/areas"
          }
        },
        "backlog": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/backlog"
          }
        },
        "codebases": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/codebases"
          }
        },
        "collaborators": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/collaborators"
          }
        },
        "filters": {
          "links": {
            "related": "http://minishift.local:30000/api/filters"
          }
        },
        "iterations": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/iterations"
          }
        },
        "labels": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/labels"
          }
        },
        "owned-by": {
          "data": {
            "id": "2dd07975-e8c5-4017-9003-b3f34aa32028",
            "type": "identities"
          },
          "links": {
            "related": "http://minishift.local:30000/api/users/2dd07975-e8c5-4017-9003-b3f34aa32028"
          }
        },
        "space-template": {
          "data": {
            "id": "929c963a-174c-4c37-b487-272067e88bd4",
            "type": "spacetemplates"
          },
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4",
            "self": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4"
          }
        },
        "workitemlinktypes": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemlinktypes"
          }
        },
        "workitems": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/a36c3667-2877-4ef0-a745-524917fa6138/workitems"
          }
        },
        "workitemtypegroups": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypegroups"
          }
        },
        "workitemtypes": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypes"
          }
        }
      },
      "type": "spaces"
    },
    {
      "attributes": {
        "created-at": "2018-07-09T10:39:11.308566Z",
        "description": "This is the foobar collaboration space",
        "name": "space_2",
        "updated-at": "2018-07-09T10:39:11.308566Z",
        "version": 0
      },
      "id": "1b6be181-7f67-4e9b-a6f4-8588565df4bd",
      "links": {
        "backlog": {
          "self": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/backlog"
        },
        "filters": "http://minishift.local:30000/api/filters",
        "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd",
        "self": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd",
        "workitemlinktypes": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemlinktypes",
        "workitemtypes": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypes"
      },
      "relationships": {
        "areas": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/areas"
          }
        },
        "backlog": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/backlog"
          }
        },
        "codebases": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/codebases"
          }
        },
        "collaborators": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/collaborators"
          }
        },
        "filters": {
          "links": {
            "related": "http://minishift.local:30000/api/filters"
          }
        },
        "iterations": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/iterations"
          }
        },
        "labels": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/labels"
          }
        },
        "owned-by": {
          "data": {
            "id": "2dd07975-e8c5-4017-9003-b3f34aa32028",
            "type": "identities"
          },
          "links": {
            "related": "http://minishift.local:30000/api/users/2dd07975-e8c5-4017-9003-b3f34aa32028"
          }
        },
        "space-template": {
          "data": {
            "id": "929c963a-174c-4c37-b487-272067e88bd4",
            "type": "spacetemplates"
          },
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4",
            "self": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4"
          }
        },
        "workitemlinktypes": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemlinktypes"
          }
        },
        "workitems": {
          "links": {
            "related": "http://minishift.local:30000/api/spaces/1b6be181-7f67-4e9b-a6f4-8588565df4bd/workitems"
          }
        },
        "workitemtypegroups": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypegroups"
          }
        },
        "workitemtypes": {
          "links": {
            "related": "http://minishift.local:30000/api/spacetemplates/929c963a-174c-4c37-b487-272067e88bd4/workitemtypes"
          }
        }
      },
      "type": "spaces"
    }
  ],
  "links": {
    "first": "http://minishift.local:30000/api/search/codebases?page[offset]=0&page[limit]=20&url=https://github.com/surajssd/heyheyworld",
    "last": "http://minishift.local:30000/api/search/codebases?page[offset]=0&page[limit]=20&url=https://github.com/surajssd/heyheyworld"
  },
  "meta": {
    "totalCount": 2
  }
}
```
</details>